README.md
# Delethon

<escape><a href="https://travis-ci.org/BingLingGroup/delethon"><img src="https://travis-ci.org/BingLingGroup/delethon.svg?branch=dev"></img></a></escape>

[简体中文](docs/README.zh-Hans.md)

[Changelog](CHANGELOG.md)

<escape><img src="docs/icon/delethon.png" width="64px"></escape>

Icon font: [source-han-serif](https://source.typekit.com/source-han-serif/) ([OFL 1.1](https://github.com/adobe-fonts/source-han-serif/blob/master/LICENSE.txt))

### TOC

1. [Description](#description)
2. [License](#license)
3. [Dependencies](#dependencies)
4. [Download and Installation](#download-and-installation)
   - 4.1 [Branches](#branches)
   - 4.2 [Install on Ubuntu](#install-on-ubuntu)
   - 4.3 [Install on Windows](#install-on-windows)
5. [Usage](#usage)
   - 6.1 [Typical usage](#typical-usage)
   - 6.2 [Options](#Options)
   - 6.3 [Internationalization](#internationalization)
6. [Bugs report](#bugs-report)
7. [Build](#build)

Click up arrow to go back to TOC.

### Description

Delete Telegram messages based on Telethon.

It can do:

- [x] Delete the messages of the logged in user for all members in group chats.
- [x] Delete certain types of the messages.
- [x] Print instead of deleting the messages.

It can not do:

- [ ] Delete the messages that's been cached, [anti-revoked](https://github.com/SpriteOvO/Telegram-Anti-Revoke) or exported to the local storage.
- [ ] Delete the messages in group's recent actions.
- [ ] Delete other people's messages for all in group chats unless you have the admin rights of the group.

### License

[GPLv3](LICENSE)

### Dependencies

[requirements.txt](requirements.txt)

Python >= 3.5

### Download and Installation

About the git installation. If you don't want to install git to use pip [VCS](https://pip.pypa.io/en/stable/reference/pip_install/#vcs-support) support to install python package or just confused with git environment variables, you can manually click that clone and download button to download the source code and use pip to install the source code [locally](https://pip.pypa.io/en/stable/reference/pip_install/#description) by input these commands.

```batch
cd the_directory_contains_the_source_code
pip install .
```

#### Branches

[master branch](https://github.com/BingLingGroup/delethon/tree/master)

- Codes will be updated when a new version releases.

[dev branch](https://github.com/BingLingGroup/delethon/tree/dev)

- The latest codes will be pushed to this branch. It will be merged to master branch when new version released.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

#### Install on Ubuntu

Include dependencies installation commands.

Install from `dev` branch.

```bash
apt install python3 python3-pip git -y
pip3 install git+https://github.com/BingLingGroup/delethon.git@dev
```

Install from `master` branch.

```bash
apt install python3 python3-pip git -y
pip3 install git+https://github.com/BingLingGroup/delethon.git@master
```

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

#### Install on Windows

You can just go to the [release page](https://github.com/BingLingGroup/delethon/releases) and download the latest release for Windows. The click-and-run batches are also in the package. You can manually edit by using Notepad++. Or add the executable files' directory to system environment variables so you can use it as a universal command everywhere in the system if permission is Ok.

Tips: `Shift - Right Click` is the keyboard shortcut for opening a Powershell on current directory. To open an exe at current directory, the format is like `.\delethon`.

Or install Python environment(if you still don't have one) from [choco](https://chocolatey.org) and then install the package.

Choco installation command is for cmd.(not Powershell)

```batch
@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))" && SET "PATH=%PATH%;%ALLUSERSPROFILE%\chocolatey\bin"
```

Install from `dev` branch.

```batch
choco install git python curl -y
curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
python get-pip.py
pip install git+https://github.com/BingLingGroup/delethon.git@dev
```

Install from `master` branch.

```batch
choco install git python curl -y
curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
python get-pip.py
pip install git+https://github.com/BingLingGroup/delethon.git@master
```

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

### Usage

You need to get the `api_id` and `api_hash` from [Telegram app](https://my.telegram.org/apps) to log in. Remember not to post these codes anywhere or give them to anyone.

If the user is logged in for the first time or the session file is not ready, it will ask for a login authentication just like other Telegram clients.

Typical usages for delethon instead of other Telegram clients:

- Delete all messages of the logged in user in group chats instead of asking their admin's for help.
- Delete all messages of the group chats on schedule if the logged in user is the admin of some group chats.

Log in behind local `SOCKS5` proxy `127.0.0.1:1080`

```
delethon -pt ...(other options)
```

Delete all messages of logged in user in some chats.

```
delethon -ai api_id -ah api_hash -c "chat_1" "chat_2" -m ...(other options)
```

Delete quietly.

```
delethon -ll 0 ...(other options)
```

Delete text messages only. (But won't delete those photo messages with caption)

```
delethon -fs empty ...(other options)
```

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

#### Options

Full list of the help message.

```
$ delethon -h
usage:
  delethon [options]

Delete Telegram messages based on Telethon.

Client Options:
  Options to start a Telethon client.

  -ai API_ID, --api-id API_ID
                        Telegram app api_id. Input it or set the environment
                        variable "TELEGRAM_API_ID". You can get one from
                        https://my.telegram.org/apps (arg_num = 1) (default:
                        None)
  -ah API_HASH, --api-hash API_HASH
                        Telegram app api_hash. Input it or set the environment
                        variable "TELEGRAM_API_HASH". You can get one from
                        https://my.telegram.org/apps (arg_num = 1) (default:
                        None)
  -sf path, --session-file path
                        Telegram session file path. Ref: https://docs.telethon
                        .dev/en/latest/concepts/sessions.html (arg_num = 1)
                        (default: Telethon)

Proxy Options:
  Options to use Telethon behind a proxy.

  -pt [proxy_type], --proxy-type [proxy_type]
                        PySocks or MTProto Proxy. Available types: "SOCKS5",
                        "SOCKS4", "HTTP", "MTPROTO". When using "MTPROTO", its
                        type is "ConnectionTcpMTProxyRandomizedIntermediate".
                        If environment variable "HTTP_PROXY" exists and this
                        option is not used, it will use it. Ref:
                        https://docs.telethon.dev/en/latest/basic/signing-
                        in.html#signing-in-behind-a-proxy If arg_num is 0, use
                        const type. (arg_num = 0 or 1) (const: SOCKS5)
  -pa address, --proxy-address address
                        The IP address or DNS name of the proxy server.
                        (arg_num = 1) (default: 127.0.0.1)
  -pp port, --proxy-port port
                        The port of the proxy server. (arg_num = 1) (default:
                        1080)
  -pu username, --proxy-username username
                        Set proxy username. (arg_num = 1)
  -ps password, --proxy-password password
                        Set proxy password. When using "MTPROTO", this option
                        is "secret" instead. (arg_num = 1)

Iterate Messages Options:
  Options to control Telethon iter_messages method. Ref: https://docs.telethon.dev/en/latest/modules/client.html#telethon.client.messages.MessageMethods.iter_messages

  -c [entity_like [entity_like ...]], --chats [entity_like [entity_like ...]]
                        Get the messages in the chat entity/entities.
                        entity_like can be Telegram username, Group name and
                        so on. If failed to get one, it will use it to match
                        the entity name or id of the dialogs which the user
                        joined. Ref: https://docs.telethon.dev/en/latest/conce
                        pts/entities.html#getting-entities(arg_num >= 1)
  -u [entity_like [entity_like ...]], --users [entity_like [entity_like ...]]
                        Get the messages from the user entity/entities. If not
                        input, gets all users messages. Ref: https://docs.tele
                        thon.dev/en/latest/concepts/entities.html#getting-
                        entities(arg_num >= 1)
  -m, --me              Get messages from the current user who is logged in.
                        (arg_num = 0)
  -ac, --all-chats      Iterate over all the chat entity/entities that the
                        current user joined. (arg_num = 0)
  -l int, --limit int   Number of messages to be retrieved. Slower when more
                        than 3000. Ref: https://docs.telethon.dev/en/latest/mo
                        dules/client.html#telethon.client.messages.MessageMeth
                        ods.iter_messages (arg_num = 1) (default: unlimited)
  -ofd int, --offset-day int
                        Offset day (messages previous to this day will be
                        retrieved). Exclusive. (arg_num = 1)
  -ofi int, --offset-id int
                        Offset message ID (only messages previous to the given
                        ID will be retrieved). Exclusive. (arg_num = 1)
  -mxi int, --max-id int
                        All the messages with a higher (newer) ID or equal to
                        this will be excluded. (arg_num = 1)
  -mni int, --min-id int
                        All the messages with a lower (older) ID or equal to
                        this will be excluded. (arg_num = 1)
  -aof int, --add-offset int
                        Additional message offset (all of the specified
                        offsets + this offset = older messages). (arg_num = 1)
  -s str, --search str  The string to be used as a search query. Give the same
                        result as other Telegram official clients meaning it's
                        not optimized for some non-English languages. (arg_num
                        = 1)
  -f type, --filter type
                        The filter to use before returning messages. For
                        instance, "photos" for "InputMessagesFilterPhotos"
                        would yield only messages containing photos. When
                        using "empty" filter, you must give option "--search".
                        Available MessagesFilters: chatphotos, contacts,
                        document, empty, geo, gif, music, mentions,
                        phonecalls, photovideo, photos, roundvideo,
                        roundvoice, url, video, voice (arg_num = 1)
  -wt int, --wait-time int
                        Wait time (in seconds) between different
                        GetHistoryRequest. Use this parameter to avoid hitting
                        the FloodWaitError as needed. If left to None, it will
                        default to 1 second only if the limit is higher than
                        3000. If the ids parameter is used, this time will
                        default to 10 seconds only if the amount of IDs is
                        higher than 300. (arg_num = 1)
  -ids [int [int ...]], --ids [int [int ...]]
                        A single integer ID (or several IDs) for the message
                        that should be returned. This parameter takes
                        precedence over the rest (which will be ignored if
                        this is set). This can for instance be used to get the
                        message with ID 123 from a channel. Note that if the
                        message doesn't exist, None will appear in its place,
                        so that zipping the list of IDs with the messages can
                        match one-to-one. (arg_num >= 1)
  -r, --reverse         The messages will be returned in reverse order (from
                        oldest to newest, instead of the default newest to
                        oldest). This also means that the meaning of "--
                        offset-id" and "--offset-day" parameters is reversed,
                        although they will still be exclusive. "--min-id"
                        becomes equivalent to "--offset-id" instead of being "
                        --max-id" as well since messages are returned in
                        ascending order. (arg_num = 0)

Process Options:
  Options to determine how to process the messages after retrieving them.

  -op, --only-print     Only print message instead of deleting them. (arg_num
                        = 0)
  -fs [type [type ...]], --filters [type [type ...]]
                        The filters to use after returning the messages from
                        "iter_messages". So using it will delete equal to or
                        less amount of messages than the argument set to "--
                        limit". It run slower than option "--filter", but
                        supports multiple filters. And it support "empty" type
                        without input "--search". Available MessageMedia
                        filters: contact, document, empty, game, geo, geolive,
                        invoice, photo, poll, unsupported, venue, webpage Ref:
                        https://tl.telethon.dev/types/message_media.html(arg_n
                        um >= 1)

Information Options:
  Options to get extra information.

  -ll int, --log-level int
                        Print different kinds of messages. 0 for nothing. 1
                        for basic. 2 for more content. (arg_num = 1) (default:
                        2)
  -h, --help            Show delethon help message and exit. (arg_num = 0)
  -v, --version         Show delethon version and exit. (arg_num = 0)

Make sure the argument with space is in quotes.
The default value is used
when the option is not given at the command line.
"(arg_num)" means if the option is given,
the number of the arguments is required.
Author: Bing Ling
Email: binglinggroup@outlook.com
Bug report: https://github.com/BingLingGroup/delethon
```

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

#### Internationalization

Delethon supports multi-language command line user interface by [GNU gettext](https://www.gnu.org/software/gettext/). Now supports `zh_CN` and default `en_US`. More info about this [lang codes format](https://www.gnu.org/software/gettext/manual/gettext.html#Locale-Names). The program will automatically detect the os locale and use the one supported. For windows 10, it seems adjusting the `Region`-`Regional format` is Ok.

Of course, delethon offers a method to override the os locale. Just create a txt file without extension named `locale`, containing the lang codes at the beginning of the file, at the command line current working directory. When delethon starts, it will detect this file and read the lang code inside it and apply it if supported.

If you want to translate this program into other languages, first install the gettext utilities. Then you can run `python scripts/update_po_files.py lang_code` to create the locale files which you want to translate into. And then use [POEditor](https://poeditor.com/) to edit po files. [update_po_files.py](scripts/update_po_files.py) can also automatically merge the position info into the old po files and compile the po files into mo files which the program read them. So it's useful when the codes changed, you can merge the positional changes into the translations automatically.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

### Bugs report

Bugs and suggestions can be reported at [issues](https://github.com/BingLingGroup/delethon/issues).

### Build

Current windows build is built by [pyinstaller script](scripts/pyinstaller_build.bat).

[create_release.py](scripts/create_release.py) is used to make the release package.


CHANGELOG.md
## Changelog

[简体中文](docs/CHANGELOG.zh-Hans.md)

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## TOC

- [Unreleased](#unreleased)
- [0.1.1-alpha - 2020-03-18](#011-alpha---2020-03-18)
  - [Added](#added011-alpha)
  - [Changed](#changed011-alpha)
- [0.1.0-alpha - 2020-03-11](#010-alpha---2020-03-11)
  - [Added](#added010-alpha)

Click up arrow to go back to TOC.

### [Unreleased]

### [0.1.1-alpha] - 2020-03-18

#### Added(0.1.1-alpha)

- Add option to input a group of options to avoid starting the delethon multiple times.

#### Changed(0.1.1-alpha)

- Fix print bugs.

### [0.1.0-alpha] - 2020-03-11

#### Added(0.1.0-alpha)

- Add basic messages deleting functions.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

[Unreleased]: https://github.com/BingLingGroup/autosub/compare/0.1.1-alpha...HEAD
[0.1.1-alpha]: https://github.com/BingLingGroup/autosub/compare/0.1.0-alpha...0.1.1-alpha
[0.1.0-alpha]: https://github.com/BingLingGroup/autosub/releases/tag/0.1.0-alpha


delethon/options.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Defines delethon's options.
"""

# Import built-in modules
import argparse
import gettext
import os

# Import third-party modules


# Any changes to the path and your own modules
from delethon import metadata
from delethon import constants


OPTIONS_TEXT = gettext.translation(domain=__name__,
                                   localedir=constants.LOCALE_PATH,
                                   languages=[constants.CURRENT_LOCALE],
                                   fallback=True)

META_TEXT = gettext.translation(domain=metadata.__name__,
                                localedir=constants.LOCALE_PATH,
                                languages=[constants.CURRENT_LOCALE],
                                fallback=True)

_ = OPTIONS_TEXT.gettext
M_ = META_TEXT.gettext


def get_cmd_parser():
    """
    Get command-line arguments.
    """
    parser = argparse.ArgumentParser(
        prog=metadata.NAME,
        usage=_('\n  %(prog)s [options]'),
        description=M_(metadata.DESCRIPTION),
        epilog=_("Make sure the argument with space is in quotes.\n"
                 "The default value is used\n"
                 "when the option is not given at the command line.\n"
                 "\"(arg_num)\" means if the option is given,\n"
                 "the number of the arguments is required.\n"
                 "Author: {author}\n"
                 "Email: {email}\n"
                 "Bug report: {homepage}\n").format(
                     author=metadata.AUTHOR,
                     email=metadata.AUTHOR_EMAIL,
                     homepage=metadata.HOMEPAGE),
        add_help=False,
        formatter_class=argparse.RawDescriptionHelpFormatter)

    client_group = parser.add_argument_group(
        _('Client Options'),
        _('Options to start a Telethon client.'))

    proxy_group = parser.add_argument_group(
        _('Proxy Options'),
        _('Options to use Telethon behind a proxy.')
    )

    iter_messages_group = parser.add_argument_group(
        _('Iterate Messages Options'),
        _('Options to control Telethon iter_messages method. '
          'Ref: https://docs.telethon.dev/en/latest/modules'
          '/client.html#telethon.client.messages.MessageMethods.iter_messages')
    )

    process_group = parser.add_argument_group(
        _('Process Options'),
        _('Options to determine how to process the messages after retrieving them.')
    )

    info_group = parser.add_argument_group(
        _('Information Options'),
        _('Options to get extra information.')
    )

    options_group = parser.add_argument_group(
        _('Group Options'),
        _('Options to input group of arguments.')
    )

    client_group.add_argument(
        '-ai', '--api-id',
        metavar='API_ID',
        type=int,
        default=os.environ.get('TELEGRAM_API_ID'),
        help=_("Telegram app api_id. Input it or set the environment variable "
               "\"TELEGRAM_API_ID\". "
               "You can get one from https://my.telegram.org/apps "
               "(arg_num = 1) (default: %(default)s)")
    )

    client_group.add_argument(
        '-ah', '--api-hash',
        metavar='API_HASH',
        default=os.environ.get('TELEGRAM_API_HASH'),
        help=_("Telegram app api_hash. Input it or set the environment variable "
               "\"TELEGRAM_API_HASH\". "
               "You can get one from https://my.telegram.org/apps "
               "(arg_num = 1) (default: %(default)s)")
    )

    client_group.add_argument(
        '-sf', '--session-file',
        metavar=_('path'),
        default='Telethon',
        help=_("Telegram session file path. "
               "Ref: https://docs.telethon.dev/en/latest/concepts/sessions.html "
               "(arg_num = 1) (default: %(default)s)")
    )

    proxy_group.add_argument(
        '-pt', '--proxy-type',
        metavar=_('proxy_type'),
        nargs='?',
        const='SOCKS5',
        help=_("PySocks or MTProto Proxy. "
               "Available types: \"SOCKS5\", \"SOCKS4\", "
               "\"HTTP\", \"MTPROTO\". "
               "When using \"MTPROTO\", its type is "
               "\"ConnectionTcpMTProxyRandomizedIntermediate\". "
               "If environment variable \"HTTP_PROXY\" exists "
               "and this option is not used, "
               "it will use it. "
               "Ref: https://docs.telethon.dev/en/latest/basic"
               "/signing-in.html#signing-in-behind-a-proxy "
               "If arg_num is 0, use const type. "
               "(arg_num = 0 or 1) (const: %(const)s)")
    )

    proxy_group.add_argument(
        '-pa', '--proxy-address',
        metavar=_('address'),
        default='127.0.0.1',
        help=_("The IP address or DNS name of the proxy server. "
               "(arg_num = 1) (default: %(default)s)")
    )

    proxy_group.add_argument(
        '-pp', '--proxy-port',
        metavar=_('port'),
        default=1080,
        type=int,
        help=_("The port of the proxy server. "
               "(arg_num = 1) (default: %(default)s)")
    )

    proxy_group.add_argument(
        '-pu', '--proxy-username',
        metavar=_('username'),
        help=_("Set proxy username. "
               "(arg_num = 1)")
    )

    proxy_group.add_argument(
        '-ps', '--proxy-password',
        metavar=_('password'),
        help=_("Set proxy password. "
               "When using \"MTPROTO\", this option is \"secret\" instead. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-c', '--chats',
        nargs='*',
        metavar=_('entity_like'),
        help=_("Get the messages in the chat entity/entities. "
               "entity_like can be Telegram username, Group name and so on. "
               "If failed to get one, it will use it to match the entity name or id "
               "of the dialogs which the user joined. "
               "Ref: https://docs.telethon.dev/en/"
               "latest/concepts/entities.html#getting-entities"
               "(arg_num >= 1)")
    )

    iter_messages_group.add_argument(
        '-u', '--users',
        nargs='*',
        metavar=_('entity_like'),
        help=_("Get the messages from the user entity/entities. "
               "If not input, gets all users messages. "
               "Ref: https://docs.telethon.dev/en/"
               "latest/concepts/entities.html#getting-entities"
               "(arg_num >= 1)")
    )

    iter_messages_group.add_argument(
        '-m', '--me',
        action='store_true',
        help=_("Get messages from the current user who is logged in. "
               "(arg_num = 0)"))

    iter_messages_group.add_argument(
        '-ac', '--all-chats',
        action='store_true',
        help=_("Iterate over all the chat entity/entities "
               "that the current user joined. "
               "(arg_num = 0)"))

    iter_messages_group.add_argument(
        '-l', '--limit',
        metavar=_('int'),
        type=int,
        help=_("Number of messages to be retrieved. "
               "Slower when more than 3000. "
               "Ref: https://docs.telethon.dev/en/latest/modules"
               "/client.html#telethon.client.messages.MessageMethods.iter_messages "
               "(arg_num = 1) (default: unlimited)")
    )

    iter_messages_group.add_argument(
        '-ofd', '--offset-day',
        metavar=_('int'),
        type=int,
        help=_("Offset day (messages previous to this day will be retrieved). "
               "Exclusive. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-ofi', '--offset-id',
        metavar=_('int'),
        type=int,
        default=0,
        help=_("Offset message ID (only messages previous to the given ID "
               "will be retrieved). "
               "Exclusive. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-mxi', '--max-id',
        metavar=_('int'),
        type=int,
        default=0,
        help=_("All the messages with a higher (newer) ID "
               "or equal to this will be excluded. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-mni', '--min-id',
        metavar=_('int'),
        type=int,
        default=0,
        help=_("All the messages with a lower (older) ID "
               "or equal to this will be excluded. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-aof', '--add-offset',
        metavar=_('int'),
        type=int,
        default=0,
        help=_("Additional message offset "
               "(all of the specified offsets + this offset = older messages). "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-s', '--search',
        metavar=_('str'),
        help=_("The string to be used as a search query. "
               "Give the same result as other Telegram official clients "
               "meaning it's not optimized for some non-English languages. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-f', '--filter',
        metavar=_('type'),
        help=_("The filter to use before returning messages. "
               "For instance, \"photos\" for \"InputMessagesFilterPhotos\" "
               "would yield only messages containing photos. "
               "When using \"empty\" filter, you must give option \"--search\". "
               "Available MessagesFilters: chatphotos, contacts, document, "
               "empty, geo, gif, music, mentions, phonecalls, photovideo, "
               "photos, roundvideo, roundvoice, url, video, voice "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-wt', '--wait-time',
        metavar=_('int'),
        type=int,
        help=_("Wait time (in seconds) between different GetHistoryRequest. "
               "Use this parameter to avoid hitting the FloodWaitError as needed. "
               "If left to None, it will default to 1 second "
               "only if the limit is higher than 3000. "
               "If the ids parameter is used, "
               "this time will default to 10 seconds only "
               "if the amount of IDs is higher than 300. "
               "(arg_num = 1)")
    )

    iter_messages_group.add_argument(
        '-ids', '--ids',
        metavar=_('int'),
        type=int,
        nargs='*',
        help=_("A single integer ID (or several IDs) for the message "
               "that should be returned. This parameter takes precedence "
               "over the rest (which will be ignored if this is set). "
               "This can for instance be used to get the message "
               "with ID 123 from a channel. Note that if the message "
               "doesn't exist, None will appear in its place, "
               "so that zipping the list of IDs with the messages "
               "can match one-to-one. "
               "(arg_num >= 1)")
    )

    iter_messages_group.add_argument(
        '-r', '--reverse',
        action='store_true',
        help=_("The messages will be returned in reverse order"
               " (from oldest to newest, instead of the default newest to oldest). "
               "This also means that the meaning of \"--offset-id\" and \"--offset-day\" "
               "parameters is reversed, although they will still be exclusive. "
               "\"--min-id\" becomes equivalent to \"--offset-id\" "
               "instead of being \"--max-id\" as "
               "well since messages are returned in ascending order. "
               "(arg_num = 0)"))

    process_group.add_argument(
        '-op', '--only-print',
        action='store_true',
        help=_("Only print message instead of deleting them. "
               "(arg_num = 0)"))

    process_group.add_argument(
        '-fs', '--filters',
        nargs='*',
        metavar=_('type'),
        help=_("The filters to use after returning the messages from \"iter_messages\". "
               "So using it will delete equal to or less amount of messages than "
               "the argument set to \"--limit\". "
               "It run slower than option \"--filter\", but supports multiple filters. "
               "And it support \"empty\" type without input \"--search\". "
               "Available MessageMedia filters: contact, document, empty, "
               "game, geo, geolive, invoice, photo, poll, unsupported, "
               "venue, webpage "
               "Ref: https://tl.telethon.dev/types/message_media.html"
               "(arg_num >= 1)")
    )

    info_group.add_argument(
        '-ll', '--log-level',
        metavar=_('int'),
        type=int,
        default=2,
        choices=range(1, 3),
        help=_("Print different kinds of messages. "
               "0 for nothing. 1 for basic. 2 for more content."
               " (arg_num = 1) (default: %(default)s)"))

    info_group.add_argument(
        '-h', '--help',
        action='help',
        help=_("Show %(prog)s help message and exit. (arg_num = 0)"))

    info_group.add_argument(
        '-v', '--version',
        action='version',
        version='%(prog)s ' + metadata.VERSION
                + ' by ' + metadata.AUTHOR + ' <'
                + metadata.AUTHOR_EMAIL + '>',
        help=_("Show %(prog)s version and exit. (arg_num = 0)"))

    options_group.add_argument(
        '-a', '--argv',
        nargs=argparse.REMAINDER,
        metavar=_('option_group'),
        help=_("A group of options to input to "
               "avoid starting the delethon multiple times. "
               "Except for the first group, "
               "if the group options don't include \"-ai\" or \"-ah\", "
               "it will avoid starting the client multiple times "
               "to avoid extra consumption. "
               "Remember to input the group of options in quotes "
               "and escape the original quotes by using backslash \"\\\". "
               "(arg_num >= 1)")
    )

    return parser
